File: test/phate_examples.py
import phate
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import sys


def main():
    python_version = "{}.{}".format(
        sys.version_info.major, sys.version_info.minor)

    # generate DLA tree
    M, C = phate.tree.gen_dla(n_dim=100, n_branch=10, branch_length=300,
                              rand_multiplier=2, seed=37, sigma=4)

    # instantiate phate_operator
    phate_operator = phate.PHATE(n_components=2, a=10, k=5, t=30, mds='classic',
                                 knn_dist='euclidean', mds_dist='euclidean',
                                 njobs=-2, n_landmark=None)

    # run phate with classic MDS
    print("DLA tree, classic MDS")
    Y_cmds = phate_operator.fit_transform(M)

    # run phate with metric MDS
    # change the MDS embedding without recalculating diffusion potential
    phate_operator.reset_mds(mds="metric")
    print("DLA tree, metric MDS (log)")
    Y_mmds = phate_operator.fit_transform(M)

    # run phate with nonmetric MDS
    phate_operator.reset_potential(potential_method="sqrt")
    print("DLA tree, metric MDS (sqrt)")
    Y_sqrt = phate_operator.fit_transform(M)

    phate_fast_operator = phate.PHATE(
        n_components=2, a=10, t=90, k=5, mds='classic', mds_dist='euclidean',
        n_landmark=1000)
    # run phate with classic MDS
    print("DLA tree, fast classic MDS")
    Y_cmds_fast = phate_fast_operator.fit_transform(M)

    # run phate with metric MDS
    # change the MDS embedding without recalculating diffusion potential
    phate_fast_operator.reset_mds(mds="metric")
    print("DLA tree, fast metric MDS (log)")
    Y_mmds_fast = phate_fast_operator.fit_transform(M)

    # run phate with nonmetric MDS
    phate_fast_operator.reset_potential(potential_method="sqrt")
    print("DLA tree, fast metric MDS (sqrt)")
    Y_sqrt_fast = phate_fast_operator.fit_transform(M)

    fig, axes = plt.subplots(2, 3)

    (ax1, ax2, ax3) = axes[0]
    fig.set_size_inches(12, 8)
    plt.setp((ax1, ax2, ax3), xticks=[], xticklabels=[], yticks=[])

    # plotting PHATE - classic MDS
    ax1.scatter(Y_cmds[:, 0], Y_cmds[:, 1], s=10, c=C)
    ax1.set_xlabel("phate 1")
    ax1.set_ylabel("phate 2")
    ax1.set_title("PHATE embedding of DLA fractal tree\nClassic MDS")

    # plotting PHATE - metric MDS
    ax2.scatter(Y_mmds[:, 0], Y_mmds[:, 1], s=10, c=C)
    ax2.set_xlabel("phate 1")
    ax2.set_title("PHATE embedding of DLA fractal tree\nMetric MDS, log")

    # plotting PHATE - nonmetric MDS
    ax3.scatter(Y_sqrt[:, 0], Y_sqrt[:, 1], s=10, c=C)
    ax3.set_xlabel("phate 1")
    ax3.set_title("PHATE embedding of DLA fractal tree\nMetric MDS, sqrt")

    (ax1, ax2, ax3) = axes[1]
    plt.setp((ax1, ax2, ax3), xticks=[], xticklabels=[], yticks=[])

    # plotting PHATE - classic MDS
    ax1.scatter(Y_cmds_fast[:, 0], Y_cmds_fast[:, 1], s=10, c=C)
    ax1.set_xlabel("phate 1")
    ax1.set_ylabel("phate 2")
    ax1.set_title("PHATE embedding of DLA fractal tree\nFast Classic MDS")

    # plotting PHATE - metric MDS
    ax2.scatter(Y_mmds_fast[:, 0], Y_mmds_fast[:, 1], s=10, c=C)
    ax2.set_xlabel("phate 1")
    ax2.set_title("PHATE embedding of DLA fractal tree\nFast Metric MDS, log")

    # plotting PHATE - nonmetric MDS
    ax3.scatter(Y_sqrt_fast[:, 0], Y_sqrt_fast[:, 1], s=10, c=C)
    ax3.set_xlabel("phate 1")
    ax3.set_title("PHATE embedding of DLA fractal tree\nFast Metric MDS, sqrt")

    plt.tight_layout()
    plt.savefig("python{}_tree.png".format(python_version), dpi=100)

    clusters = pd.read_csv("../data/MAP.csv", header=None)
    clusters.columns = pd.Index(['wells', 'clusters'])
    bmmsc = pd.read_csv("../data/BMMC_myeloid.csv.gz", index_col=0)

    C = clusters['clusters']  # using cluster labels from original publication

    # library_size_normalize performs L1 normalization on each cell
    bmmsc_norm = phate.preprocessing.library_size_normalize(bmmsc)
    phate_operator = phate.PHATE(
        n_components=2, t='auto', a=200, k=10, mds='metric', mds_dist='euclidean',
        n_landmark=None)
    phate_fast_operator = phate.PHATE(
        n_components=2, t='auto', a=200, k=10, mds='metric', mds_dist='euclidean',
        n_landmark=1000)

    fig, (ax1, ax2) = plt.subplots(1, 2, sharey='all')
    print("BMMSC, exact PHATE")
    Y_mmds = phate_operator.fit_transform(bmmsc_norm, t_max=100,
                                          plot_optimal_t=True, ax=ax1)
    ax1.set_title("Optimal t selection on 2730 BMMSCs\n"
                  "Exact PHATE, " + ax1.get_title())
    print("BMMSC, fast PHATE")
    Y_mmds_fast = phate_fast_operator.fit_transform(bmmsc_norm, t_max=100,
                                                    plot_optimal_t=True, ax=ax2)
    ax2.set_title("Optimal t selection on 2730 BMMSCs\n"
                  "Fast PHATE, " + ax2.get_title())
    plt.tight_layout()
    plt.savefig("python{}_bmmsc_optimal_t.png".format(python_version), dpi=100)

    fig, (ax1, ax2) = plt.subplots(1, 2)
    plt.setp((ax1, ax2), xticklabels=[], yticklabels=[])

    ax1.scatter(Y_mmds[:, 0], Y_mmds[:, 1], s=1, c=C, cmap="Paired")
    ax1.set_xlabel("phate 1")
    ax1.set_ylabel("phate 2")
    ax1.set_title("PHATE embedding of 2730 BMMSCs\nExact PHATE")

    ax2.scatter(Y_mmds_fast[:, 0], Y_mmds_fast[:, 1],
                s=1, c=C, cmap="Paired")
    ax2.set_xlabel("phate 1")
    ax2.set_ylabel("phate 2")
    ax2.set_title("PHATE embedding of 2730 BMMSCs\nFast PHATE")

    plt.gcf().set_size_inches(8, 8)
    plt.tight_layout()
    plt.savefig("python{}_bmmsc.png".format(python_version), dpi=100)
    return 0

if __name__ == "__main__":
    exit(main())

File: test/phate_examples.R
library(phateR)
library(ggplot2)
library(gridGraphics)
library(cowplot)

# generate DLA tree

data(tree.data)
M <- tree.data$data
C <- tree.data$branches

# run phate with classic MDS
print("DLA tree, classic MDS")
Y_cmds <- phate(M, ndim=2, alpha=10, k=5, t=30, mds.method='classic',
                knn.dist.method='euclidean', mds.dist.method='euclidean',
                n.landmark=NA)

# run phate with metric MDS
print("DLA tree, metric MDS (log)")
Y_mmds <- phate(M, ndim=2, alpha=10, k=5, t=30, mds.method='metric',
                knn.dist.method='euclidean', mds.dist.method='euclidean',
                n.landmark=NA, init=Y_cmds)

# run phate with nonmetric MDS
print("DLA tree, metric MDS (sqrt)")
Y_sqrt <- phate(M, ndim=2, alpha=10, k=5, t=30, mds.method='metric',
                knn.dist.method='euclidean', mds.dist.method='euclidean',
                n.landmark=NA, potential.method='sqrt', init=Y_mmds)

# run phate with classic MDS
print("DLA tree, fast classic MDS")
Y_cmds_fast <- phate(M, ndim=2, alpha=10, k=5, t=90, mds.method='classic',
                knn.dist.method='euclidean', mds.dist.method='euclidean',
                n.landmark=1000)

# run phate with metric MDS
print("DLA tree, fast metric MDS (log)")
Y_mmds_fast <- phate(M, ndim=2, alpha=10, k=5, t=90, mds.method='metric',
                knn.dist.method='euclidean', mds.dist.method='euclidean',
                n.landmark=1000, init=Y_cmds_fast)

# run phate with nonmetric MDS
print("DLA tree, fast metric MDS (sqrt)")
Y_sqrt_fast <- phate(M, ndim=2, alpha=10, k=5, t=90, mds.method='metric',
                knn.dist.method='euclidean', mds.dist.method='euclidean',
                n.landmark=1000, potential.method='sqrt', init=Y_mmds_fast)

p <- plot_grid(ggplot(Y_cmds) + 
            geom_point(aes(PHATE1, PHATE2, color=C), show.legend=FALSE) + 
            labs(title="PHATE embedding of DLA fractal tree", 
                 subtitle="Classic MDS"),
          ggplot(Y_mmds) + 
            geom_point(aes(PHATE1, PHATE2, color=C), show.legend=FALSE) + 
            labs(title="PHATE embedding of DLA fractal tree", 
                 subtitle="Metric MDS, log"),
          ggplot(Y_sqrt) + 
            geom_point(aes(PHATE1, PHATE2, color=C), show.legend=FALSE) + 
            labs(title="PHATE embedding of DLA fractal tree", 
                 subtitle="Metric MDS, sqrt"),
          ggplot(Y_cmds_fast) + 
            geom_point(aes(PHATE1, PHATE2, color=C), show.legend=FALSE) + 
            labs(title="PHATE embedding of DLA fractal tree", 
                 subtitle="Fast classic MDS"),
          ggplot(Y_mmds_fast) + 
            geom_point(aes(PHATE1, PHATE2, color=C), show.legend=FALSE) + 
            labs(title="PHATE embedding of DLA fractal tree", 
                 subtitle="Fast metric MDS, log"),
          ggplot(Y_sqrt_fast) + 
            geom_point(aes(PHATE1, PHATE2, color=C), show.legend=FALSE) + 
            labs(title="PHATE embedding of DLA fractal tree", 
                 subtitle="Fast metric MDS, sqrt"),
          ncol=3)
save_plot("R_tree.png", p, base_height = 6, base_width=12)

clusters <- read.csv("../data/MAP.csv", row.names=NULL, header=FALSE,
                     col.names=c('wells', 'clusters'))
bmmsc <- read.csv("../data/BMMC_myeloid.csv.gz", row.names=NULL)
bmmsc <- bmmsc[,2:ncol(bmmsc)]

C <- as.factor(clusters$clusters)  # using cluster labels from original publication

# library.size.normalize performs L1 normalization on each cell
bmmsc_norm <- library.size.normalize(bmmsc)

png("tmp.png")
dev.control(displaylist="enable") 
print("BMMSC, exact PHATE")
Y_mmds <- phate(bmmsc_norm, ndim=2, t='auto', a=200, k=10, 
                mds.method='metric', mds.dist.method='euclidean',
                n.landmark=NA, plot.optimal.t = TRUE)
title(main="\n\nExact PHATE on 2730 BMMSCs")
t_exact_p <- recordPlot()
dev.off()
png("tmp.png")
dev.control(displaylist="enable")  
print("BMMSC, fast PHATE")
Y_mmds_fast <- phate(bmmsc_norm, ndim=2, t='auto', a=NA, k=10, 
                     mds.method='metric', mds.dist.method='euclidean',
                     n.landmark=1000, plot.optimal.t = TRUE)
title(main="\n\nFast PHATE on 2730 BMMSCs")
t_fast_p <- recordPlot()
dev.off()
p <- plot_grid(t_exact_p, t_fast_p, ncol=2)
save_plot("R_bmmsc_optimal_t.png", p, base_height=6, base_width=8)
file.remove("tmp.png")

p <- plot_grid(ggplot(Y_mmds) + 
                 geom_point(aes(PHATE1, PHATE2, color=C), show.legend=FALSE) + 
                 labs(title="PHATE embedding of 2730 BMMSCs", 
                      subtitle="Exact PHATE"),
               ggplot(Y_mmds_fast) + 
                 geom_point(aes(PHATE1, PHATE2, color=C), show.legend=FALSE) + 
                 labs(title="PHATE embedding of 2730 BMMSCs", 
                      subtitle="Fast PHATE"),
               ncol=2)
save_plot("R_bmmsc.png", p, base_height=8, base_width=16)
